File: THCS/2023_2024/HaNoi_2324/Bai3/TROCHOI.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <fstream>

using namespace std;

int main() {
    ifstream infile("TROCHOI.INP");
    ofstream outfile("TROCHOI.OUT");

    long long N, M;
    infile >> N >> M;

    vector<pair<long long, long long>> skills(N);
    for (int i = 0; i < N; ++i) {
        infile >> skills[i].first >> skills[i].second;  // Đọc si và ei
    }

    long long max_power = 0;
    priority_queue<pair<long long, long long>> max_heap;

    // Thêm tất cả các kỹ năng vào heap
    for (const auto& skill : skills) {
        long long si = skill.first;
        long long ei = skill.second;
        max_heap.push({si + ei, ei});  // Thêm sức mạnh lần đầu tiên
    }

    // Nâng cấp trong M phút
    for (int i = 0; i < M; ++i) {
        if (!max_heap.empty()) {
            // Lấy kỹ năng có sức mạnh tối đa
            max_power += max_heap.top().first;  // Thêm sức mạnh vào kết quả
            long long next_power = max_heap.top().second;  // Chỉ số sức mạnh cho lần sau
            max_heap.pop();  // Xóa kỹ năng này
            max_heap.push({next_power, next_power});  // Thêm lại với chỉ số mới
        }
    }

    // Ghi kết quả ra tệp
    outfile << max_power << endl;

    infile.close();
    outfile.close();
    return 0;
}

File: THCS/2023_2024/HaNoi_2324/Bai3/TROCHOI.py
import heapq

def main():
    # Đọc dữ liệu
    with open('TROCHOI.INP', 'r') as f:
        N, M = map(int, f.readline().strip().split())
        skills = []
        for _ in range(N):
            si, ei = map(int, f.readline().strip().split())
            skills.append((si, ei))
    
    # Tính sức mạnh tối đa
    max_power = 0
    max_heap = []
    
    # Thêm tất cả các kĩ năng vào heap
    for si, ei in skills:
        # Lần đầu tăng cấp
        heapq.heappush(max_heap, (-(si + ei), ei))
    
    # Nâng cấp trong M phút
    for _ in range(M):
        if max_heap:
            # Lấy kĩ năng có sức mạnh tối đa
            max_power += -max_heap[0][0]  # Thêm sức mạnh vào kết quả
            # Lần tăng cấp tiếp theo
            next_power = max_heap[0][1]  # Chỉ số sức mạnh cho lần sau
            heapq.heappop(max_heap)  # Xóa kĩ năng này
            heapq.heappush(max_heap, (-next_power, next_power))  # Thêm lại với chỉ số mới
    
    # Ghi kết quả ra tệp
    with open('TROCHOI.OUT', 'w') as f:
        f.write(str(max_power) + '\n')

if __name__ == "__main__":
    main()
